### PR TITLE
Add Subscription prober

### PR DESCRIPTION
**Pull request by devguyio**
## Proposed Changes

- Adds a prober for the dispatcher subscriptions status

### TODO
- [x] Proper manual testing
- [x] Lots of cleanup
- [ ] Adding tests
- [ ] Add hash support

### CODE REVIEW

File: pkg/channel/consolidated/reconciler/controller/controller.go
Comment (matzew) [inline pkg/channel/consolidated/reconciler/controller/controller.go:24]: `v12` ?

Comment (matzew) [reply pkg/channel/consolidated/reconciler/controller/controller.go:24]: duckv1 ?

File: pkg/channel/consolidated/reconciler/controller/controller.go
Comment (matzew) [inline pkg/channel/consolidated/reconciler/controller/controller.go:76]: eps ?

File: pkg/channel/consolidated/reconciler/controller/controller.go
Comment (matzew) [inline pkg/channel/consolidated/reconciler/controller/controller.go:81]: remove empty line

File: pkg/channel/consolidated/reconciler/controller/controller.go
Comment (matzew) [inline pkg/channel/consolidated/reconciler/controller/controller.go:99]: `PodPorts` vs `Port` ?

File: pkg/channel/consolidated/reconciler/controller/controller.go
Comment (matzew) [inline pkg/channel/consolidated/reconciler/controller/controller.go:198]: usage of `lister` ?

File: pkg/channel/consolidated/reconciler/controller/kafkachannel.go
Comment (matzew) [inline pkg/channel/consolidated/reconciler/controller/kafkachannel.go:97]: typo

File: pkg/channel/consolidated/status/status.go
Comment (matzew) [inline pkg/channel/consolidated/status/status.go:2]: update your IDE template :-)

File: pkg/channel/consolidated/status/status.go
Comment (matzew) [inline pkg/channel/consolidated/status/status.go:32]: `duckv1` ?

File: pkg/channel/consolidated/status/status.go
Comment (matzew) [inline pkg/channel/consolidated/status/status.go:104]: as said above, add some comment what that all is / why

File: pkg/channel/consolidated/status/status.go
Comment (matzew) [reply pkg/channel/consolidated/status/status.go:2]: I now see why that is :-) 


Once this is merged, let's have a GH issue that we, eventually, remove all this and go w/ the `knative/pkg` bits

File: pkg/channel/consolidated/status/status.go
Comment (matzew) [inline pkg/channel/consolidated/status/status.go:157]: `computeSubscriptionHash` ?

**Comment by devguyio** [comment]
@matzew thanks for the review, I added a TODO section to show that this still needs proper cleanup (comments, naming, tests etc) , mainly would like a feedback on the approach.

**Review by slinkydeveloper — COMMENTED** [review]
This is going in the right direction :)

**Comment by matzew** [comment]
Looking at TLS failures.  Is the prober able to access a secured instance of apache kafka ?

**Comment by matzew** [comment]
@maschmid is it worth to add your test case for this scenario to this PR, or a follow up PR ?

**Comment by maschmid** [comment]
> @maschmid is it worth to add your test case for this scenario to this PR, or a follow up PR ?
> 
> eg. link to your testcase for that?

If you mean a testcase for "event can be sent immediately after the channel's subscribtion is Ready, we already have channel_single_event_test in eventing e2e... we just need to get rid of artificial sleeps like the one in

https://github.com/knative/eventing/blob/main/test/lib/operation.go#L108-L110

File: pkg/channel/consolidated/reconciler/controller/kafkachannel_test.go
Comment (lberk) [inline pkg/channel/consolidated/reconciler/controller/kafkachannel_test.go:336]: is this TODO still valid?

File: pkg/channel/consolidated/reconciler/controller/kafkachannel.go
Comment (lberk) [inline pkg/channel/consolidated/reconciler/controller/kafkachannel.go:585]: TODO?

File: pkg/channel/consolidated/dispatcher/dispatcher.go
Comment (lberk) [inline pkg/channel/consolidated/dispatcher/dispatcher.go:303]: Do we really need to pass the `groupID`? it looks like it can be constructed by all the parameters passed to the `consumerMessageHandler{}`

File: pkg/channel/consolidated/reconciler/controller/controller.go
Comment (lberk) [inline pkg/channel/consolidated/reconciler/controller/controller.go:150]: Given the lack of e2e currently, this logical group in particular might be good to add unit tests to

Comment (devguyio) [reply pkg/channel/consolidated/reconciler/controller/controller.go:150]: do you mind elaborating more? I'm not sure exactly what should such unit tests do? when pods are deleted, that `CancelPodProbing` gets called?

Comment (lberk) [reply pkg/channel/consolidated/reconciler/controller/controller.go:150]: you can inject objects into the fake informers, and ensure that the filterfunc's are behaving as expected

Comment (devguyio) [reply pkg/channel/consolidated/reconciler/controller/controller.go:150]: I think that's pretty much covered here? https://github.com/knative/pkg/blob/main/reconciler/filter_test.go

Comment (lberk) [reply pkg/channel/consolidated/reconciler/controller/controller.go:150]: I'm talking about the specific number and type of objects we'll have before, and then after a resync given the number of specific handlers and filterfuncs we have.  This part of the diff has a lot of codecov warnings about lack of unit tests.  Is there a particular reason we're looking to avoid sufficient coverage for this block?

Comment (matzew) [reply pkg/channel/consolidated/reconciler/controller/controller.go:150]: sounds reasonable to get them covered via tests. if possible.

File: pkg/channel/consolidated/reconciler/controller/controller.go
Comment (slinkydeveloper) [inline pkg/channel/consolidated/reconciler/controller/controller.go:105]: I might be wrong here, but shouldn't we enqueue the subscription (which is the one where we want to flip the ready bit)?

File: pkg/channel/consolidated/reconciler/controller/controller.go
Comment (slinkydeveloper) [inline pkg/channel/consolidated/reconciler/controller/controller.go:150]: IMO this should live inside the prober logic itself (when listing the endpoints, it should just do the difference)

File: pkg/common/consumer/consumer_handler.go
Comment (slinkydeveloper) [inline pkg/common/consumer/consumer_handler.go:32]: The way this interface evolved is a bit weird, but it's ok we can refactor later

**Review by slinkydeveloper — COMMENTED** [review]
Small comments, all addressable later

File: pkg/channel/consolidated/reconciler/controller/kafkachannel.go
Comment (lberk) [inline pkg/channel/consolidated/reconciler/controller/kafkachannel.go:258]: we've called IsReady but not used the error return value, is this the behaviour we're looking for?

File: pkg/channel/consolidated/status/status.go
Comment (slinkydeveloper) [inline pkg/channel/consolidated/status/status.go:190]: Why computing hash? Just using `sub.UID` + `sub.Generation` should be fine and be unique

File: pkg/channel/consolidated/status/status.go
Comment (lberk) [inline pkg/channel/consolidated/status/status.go:494]: As chatted, lets add a log line here (perhaps info level?) so that we know wthat we've checked dispatcher foo, and haven't found it on that particular pod.

Comment (devguyio) [reply pkg/channel/consolidated/status/status.go:494]: while I was adding the log line, I noticed I actually have a log line just above it (before the if) that's giving all the info needed. so now I'm a bit reluctant to add yet another log line with almost same details. wdyt?

File: pkg/channel/consolidated/reconciler/controller/controller.go
Comment (devguyio) [reply pkg/channel/consolidated/reconciler/controller/controller.go:105]: our `ReconcileKind` is for KafkaChannel. So we must enqueue that to which will eventually reconcile the `spec.status.subscribers` that will end up creating a `patch` only for the desired subscription.

File: pkg/channel/consolidated/dispatcher/dispatcher.go
Comment (devguyio) [reply pkg/channel/consolidated/dispatcher/dispatcher.go:303]: it's because `consumerMessageHandler` shouldn't presume how the ConsumerGroup gets calculated. here it's like this, in KafkaSource (which reuses the `consumeMessageHandler`) does something different.

File: pkg/channel/consolidated/reconciler/controller/kafkachannel_test.go
Comment (devguyio) [reply pkg/channel/consolidated/reconciler/controller/kafkachannel_test.go:336]: I'm honestly unaware of the context of this TODO. It's been there from a much earlier change that used the Kafka Channel informer to enqueue on any update

**Comment by devguyio** [comment]
going to introduce a small change to use the set of partition numbers instead of a counter
/hold

**Comment by devguyio** [comment]
@matzew @lberk @slinkydeveloper PR is ready for review. 
I changed the probing to take into consideration the `set` of partitions. So each dispatcher now exposes which subscriptions are ready and which partitions is it dispatching for. The controller need to find the set of all partitions ready to patch the `status.subscribers`
/unhold

**Comment by matzew** [comment]
>  PR is ready for review.

@devguyio I see a lot of test failures...

**Review by devguyio — COMMENTED** [review]
/retest

**Comment by devguyio** [comment]
I'll take a look at the failing tests.

**Review by slinkydeveloper — DISMISSED** [review]
For me we're good to go!

/hold for somebody else review

**Comment by devguyio** [comment]
There's a problem that manifests when running the tests in parallel, single runs of the failed tests succeed. I'll keep digging in.

**Comment by matzew** [comment]
Not sure what changed, but before the last changes the tests did actually pass, in parallel mode too

File: pkg/channel/consolidated/status/status.go
Comment (slinkydeveloper) [inline pkg/channel/consolidated/status/status.go:398]: Error is too much, Warn is fine

**Comment by devguyio** [comment]
@matzew @slinkydeveloper there's a race that shows up when cleaning up consumer groups of deleted channels (I added that cleanup as part of this PR) So, I removed that part, and we can tackle it in a separate issue that I'll create since it's a bug that's there regardless of this PR.

**Comment by slinkydeveloper** [comment]
/test pull-knative-sandbox-eventing-kafka-integration-test-channel-consolidated

Let's try again to see if test pass consistently

**Comment by slinkydeveloper** [comment]
/retest

https://prow.knative.dev/view/gs/knative-prow/pr-logs/pull/knative-sandbox_eventing-kafka/433/pull-knative-sandbox-eventing-kafka-integration-test-channel-consolidated/1372902798915538944 some kafka source test failure

**Review by slinkydeveloper — APPROVED** [review]
/lgtm

**Comment by matzew** [comment]
/lgtm
/approve

@devguyio can we get backport for 0.20 and 0.21?

**Comment by devguyio** [comment]
Yep. Will work on it.
/unhold

**Pull request merged**
